container_api/src/container.rs

```diff
@@ -0,0 +1,233 @@
+    }
+
+    pub fn stop_all(&mut self) {
+        let _ = self.instances.iter_mut().for_each(|(id, hc)| {
+            println!("Stopping instance \"{}\"...", id);
+            match hc.stop() {
+                Ok(()) => println!("ok"),
+                Err(err) => println!("Error: {}", err),
+            }
+        });
+    }
+
+    pub fn shutdown(&mut self) {
+        self.stop_all();
+        self.instances = HashMap::new();
+    }
+
+    pub fn load_config(&mut self, config: &Configuration) -> Result<(), String> {
+        let _ = config.check_consistency()?;
+        self.shutdown();
+        config
+            .instance_ids()
+            .iter()
+            .map(|id| {
+                (
+                    id,
+                    instantiate_from_config(id, config, Box::new(Container::load_dna)),
+                )
+            })
+            .for_each(|(id, maybe_holochain)| {
+                if maybe_holochain.is_ok() {
+                    self.instances.insert(id.clone(), maybe_holochain.unwrap());
+                } else {
+                    println!(
+                        "Error while trying to create instance \"{}\": {}",
+                        id,
+                        maybe_holochain.err().unwrap()
+                    );
+                }
+            });
+        Ok(())
+    }
+
+    fn load_dna(file: &String) -> Result<Dna, HolochainError> {
+        let mut f = File::open(file)?;
+        let mut contents = String::new();
+        f.read_to_string(&mut contents)?;
+        Dna::try_from(JsonString::from(contents))
+    }
+}
+
+impl<'a> TryFrom<&'a Configuration> for Container {
+    type Error = HolochainError;
+    fn try_from(config: &'a Configuration) -> Result<Self, Self::Error> {
+        let mut container = Container::new();
+        container
+            .load_config(config)
+            .map_err(|string| HolochainError::ConfigError(string))?;
+        Ok(container)
+    }
+}
+
+type DnaLoader = Box<FnMut(&String) -> Result<Dna, HolochainError> + Send>;
+
+fn instantiate_from_config(
+    id: &String,
+    config: &Configuration,
+    mut dna_loader: DnaLoader,
+) -> Result<Holochain, String> {
+    let _ = config.check_consistency()?;
+
+    config
+        .instance_by_id(&id)
+        .ok_or(String::from("Instance not found in config"))
+        .and_then(|instance_config| {
+            let agent_config = config.agent_by_id(&instance_config.agent).unwrap();
+            let dna_config = config.dna_by_id(&instance_config.dna).unwrap();
+            let dna = dna_loader((&dna_config.file)).map_err(|_| {
+                HolochainError::ConfigError(format!(
+                    "Could not load DNA file \"{}\"",
+                    dna_config.file
+                ))
+            })?;
+
+            (instance_config.storage.storage_type == "file"
+                && instance_config.storage.path.is_some())
+                .ok_or(String::from("Only file storage supported currently"))?;
```

Comment:
`try!` might actually make this more readable

Reply:
Linter thinks, ? is cleaner: <img width="692" alt="screenshot 2018-11-12 at 00 03 11" src="https://user-images.githubusercontent.com/311749/48319450-61d3c800-e60e-11e8-9d05-9874cefa8a0e.png">